### PR TITLE
Resource manager api

### PR DESCRIPTION
**Pull request by jcrist**
This adds a new prefect api for describing tasks for managing temporary resources. The intent here is to make the following example from the prefect docs simpler and less error prone:

```python
from prefect import task, Flow

@task
def setup_spark_cluster():
    return ...

@task
def run_spark_job(cluster):
    ...

@task(trigger=prefect.triggers.always_run)
def cleanup_spark_cluster(cluster):
    ...

with Flow("Spark") as flow:
    # define data dependencies
    cluster = setup_spark_cluster()
    submitted = run_spark_job(cluster)
    cleanup_spark_cluster(cluster, upstream_tasks=[submitted])

flow.set_reference_tasks([submitted])
```

This example has three tasks:
- One for creating a spark cluster
- One that uses the spark cluster
- One that cleans up the spark cluster

To ensure that the cleanup step always runs, even if the earlier tasks fail the user needs to add an `always_run` trigger to the `cleanup_spark_cluster` task. They also need to mark all tasks that rely on the cluster as upstream tasks of the cleanup task, otherwise the cluster would shutdown too early. Finally, they need to ensure the cleanup task isn't a reference task, so the success of the flow depends on the result of the spark job, not the success of the cleanup step. This is a lot for a user to remember!

To make this simpler for users, we add a new object type `ResourceManager`, which does the above for users automatically. Fundamentally execution is still the same as the above, but users don't need to do all the fiddly bits.

The easiest way to define a resource manager is using the `prefect.resource_manager` decorator. This wraps a class with two methods:

- `setup(self) -> resource`: A `setup` method that takes no arguments and returns the created resource
- `cleanup(self, resource)`: A `cleanup` method that takes the output of `setup` and cleans up the resource

Here's an example for creating a temporary dask cluster as part of the flow:

```python
from prefect import task, resource_manager, Flow, Parameter

from dask.distributed import Client


@resource_manager(name="dask")
class Dask:
    def __init__(self, n_workers=1):
        self.n_workers = n_workers

    def setup(self):
        return Client(n_workers=self.n_workers)

    def cleanup(self, client):
        client.close()


@task
def step_one(client):
    return client.submit(lambda x: x + 1, 1).result()

@task
def step_two(client, a):
    return client.submit(lambda x: x + 1, 2).result()

@task
def step_three(client, b):
    raise ValueError("Oh NO!")


with Flow("example") as flow:
    n_workers = Parameter("n_workers")

    with Dask(n_workers=n_workers) as client:
        a = step_one(client)
        b = step_two(client, a)
        c = step_three(client, b)
```

Visualizing the resulting flow:

dask

Note that the `dask.cleanup` task is properly set to run after all the tasks created in the `Dask` context. Further, the appropriate triggers and reference tasks are automatically set, so that the `dask.cleanup` task always runs but doesn't determine the flow status. If I run the above:

<details>
<summary> Output of `flow.run(n_workers=1)` </summary>

```
[2020-07-02 23:20:52] INFO - prefect.FlowRunner | Beginning Flow run for 'example'
[2020-07-02 23:20:52] INFO - prefect.FlowRunner | Starting flow run.
[2020-07-02 23:20:52] INFO - prefect.TaskRunner | Task 'n_workers': Starting task run...
[2020-07-02 23:20:52] INFO - prefect.TaskRunner | Task 'n_workers': finished task run for task with final state: 'Success'
[2020-07-02 23:20:52] INFO - prefect.TaskRunner | Task 'dask': Starting task run...
[2020-07-02 23:20:52] INFO - prefect.TaskRunner | Task 'dask': finished task run for task with final state: 'Success'
[2020-07-02 23:20:52] INFO - prefect.TaskRunner | Task 'dask.setup': Starting task run...
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'dask.setup': finished task run for task with final state: 'Success'
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'step_one': Starting task run...
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'step_one': finished task run for task with final state: 'Success'
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'step_two': Starting task run...
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'step_two': finished task run for task with final state: 'Success'
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'step_three': Starting task run...
[2020-07-02 23:20:53] ERROR - prefect.TaskRunner | Unexpected error: ValueError('Oh NO!')
Traceback (most recent call last):
  File "/Users/jcristharif/Code/prefect/src/prefect/engine/runner.py", line 48, in inner
    new_state = method(self, state, *args, **kwargs)
  File "/Users/jcristharif/Code/prefect/src/prefect/engine/task_runner.py", line 820, in get_task_run_state
    self.task.run, timeout=self.task.timeout, **raw_inputs
  File "/Users/jcristharif/Code/prefect/src/prefect/utilities/executors.py", line 188, in timeout_handler
    return fn(*args, **kwargs)
  File "test.py", line 28, in step_three
    raise ValueError("Oh NO!")
ValueError: Oh NO!
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'step_three': finished task run for task with final state: 'Failed'
[2020-07-02 23:20:53] INFO - prefect.TaskRunner | Task 'dask.cleanup': Starting task run...
[2020-07-02 23:20:55] INFO - prefect.TaskRunner | Task 'dask.cleanup': finished task run for task with final state: 'Success'
[2020-07-02 23:20:55] INFO - prefect.FlowRunner | Flow run FAILED: some reference tasks failed.
```

</details>

This PR still needs tests and docs, but I think I'm happy with the pattern as written. Opening it up now for discussion.

- [x] adds new tests (if appropriate)
- [x] add a changelog entry in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

**Comment by jcrist** [comment]
> I think we should find a way to clean this up in a visualization sense like we do with Constants. Having a client edge to every task that uses can get a bit unwieldy

I disagree - having a clear view of what tasks use the `resource` is useful IMO. At runtime there really isn't a resource-manager concept, it's just a way of adding tasks to the flow in a convenient pattern.

> This shares a name to the slightly used agent resource manager for k8s.

I wanted to use the name `resource`/`Resource`, but then it became hard to differentiate between the class with the `setup`/`cleanup` methods and the thing returned by the `setup` method.

**Comment by jcrist** [comment]
Right now we skip the `cleanup` task if the `setup` task fails. For some code (e.g. the dask example above) this is fine - a failure to setup won't leave any lingering resources. Other resources might not work that way though - an error in the setup step might need to be handled. I'm leaning towards forcing users to handle errors in the `setup` task themselves. This is similar to `__enter__`/`__exit__` - if there is a failure in the `__enter__` method, the `__exit__` method isn't run.

File: src/prefect/core/task.py [hunk 1]
```diff
@@ -173,6 +173,9 @@ def run(self, x, y):
             Task enters a failure state
         - log_stdout (bool, optional): Toggle whether or not to send stdout messages to
             the Prefect logger. Defaults to `False`.
+        - reference_task_candidate (bool, optional): Whether this task is a candidate to
```
[the remark this answers is not in the document]
Comment (jcrist) [reply src/prefect/core/task.py:176]: > This feels odd to expose to me, as it seems to be used purely for internally enforcing that cleanup tasks are never considered reference tasks.

Currently that's all it's for, yeah. Could see it being useful elsewhere though.

I'm a bit against special-casing certain task types inside the `Flow` class, especially classes that aren't also in `prefect.core`- feels off to me to have the `Flow` know about certain types of task classes. I'm not sure if tags are the right mechanism either. Of your options, setting an attribute not via the `__init__` seems the nicest to me.

[the remark this answers is not in the document]
Comment (jcrist) [reply src/prefect/core/task.py]: I did end up going with selecting on a class - the behavior we want (cleanup tasks aren't reference tasks, but children of them are, but children of a cleanup task don't exclude a parent of a cleanup task from being a reference task) was too hard to name as a boolean flag. I don't like adding more special casing of certain task types in core components, but its self contained and can be swapped for a more generic mechanism later if needed.

**Comment by jcrist** [comment]
I believe this is ready for review. I'm not sure of the module location `prefect.tasks.resources`. The user facing apis are `ResourceManager` and `resource_manager`, the object that is being managed and  passed into user tasks is called a "resource". I'm tempted to move this into `prefect.tasks.core.resource_manager.py` instead.

**Comment by joshmeek** [comment]
@jcrist This feels more to me like it belongs as a "core" task (and now that I'm thinking of it it seems like the Secret tasks belong there as well).

**Comment by jcrist** [comment]
Moved to `prefect.tasks.core.resource_manager`.

**Comment by joshmeek** [comment]
Writing comment here for posterity even though it was discussed offline. This works really well and now enables some fun things like this:

```python
@resource_manager(name="k8s_batch_client")
class K8sBatchClient:
    def __init__(self, job_names=[]):
        self.job_names = job_names

    def setup(self):
        config.load_kube_config()
        return client.BatchV1Api()

    def cleanup(self, api_instance):
        """Clean up all jobs created by this resource"""
        for job_name in self.job_names:
            api_response = api_instance.delete_namespaced_job(
                name=job_name,
                namespace="default",
                body=client.V1DeleteOptions(
                    propagation_policy="Foreground", grace_period_seconds=5
                ),
            )
            print(f"Job deleted. status={api_response.status}")
```

One "issue" encountered was a mypy complaint:
```
"K8sBatchClient" has no attribute "__enter__"mypy(error)
"K8sBatchClient" has no attribute "__exit__"mypy(error)
```

Also not something that needs to be done now but it would be nice if we could differentiate resource_manager type tasks in the UI visualization like we do for Parameters @znicholasbrown:
image

**Comment by jcrist** [comment]
The mypy thing is a bit weird - this seems to be a bug/limitation of mypy. Switching from using `resource_manager` as a decorator to a function call fixes things:

```python
K8sBatchClient = resource_manager(K8sBatchClientResource)
```

My general distaste for mypy makes me reluctant to spend time fixing this - the type annotations provided aren't wrong, mypy incorrectly assumes that the object returned from a class decorator is still the class, which it's not. Other python libraries have similar apis, so the pattern used here isn't uncommon. Not sure what (if anything) we should do about this - the decorator notation is quite nice, mypy just can't reason about it.

**Comment by sm-Fifteen** [comment]
The documentation on resource managers should probably stress how these aren't necessarily a good fit for temporary resource of all kinds. Context managers are commonly used for connections to databases or web services, and the name "Resource" (which was also used in #2891) gives the impression that these would be a good fit for things like opening a database connection and then reusing it across multiple tasks or even something like a transaction that spans more than one task. It took me a fair bit of second guessing and reading the doc to figure out that these really are a convenience wrapper around multiple tasks, but the pickleability requirement on the result of all tasks still applies, unless you want to run into problems when switching to a Dask process runner.

That is, despite the name, [this part of the documentation still applies](https://docs.prefect.io/core/advanced_tutorials/task-guide.html#task-inputs-and-outputs):

> This subtle and technical constraint [that the data being passed between tasks must be serializable] actually informs a lot of design decisions in the Task library (as we will see shortly). Moreover, users will undoubtedly run into this situation on occasion, and it’s important to recognize the types of errors that arise when this constraint is violated. Also, as you write your own flows, keep this in mind when you design your Tasks and how they communicate information.

Using resource managers as an easy way to have volatile resources be reusable across multiple tasks looks like a simple solution based on the examples given and will actually work when using a `LocalExecutor`, but can turn into an easy footgun when attempting to then switch to parallel process execution.

**Pull request merged**